Update swsh transform utils

moxcodes opened the pull request:
## Proposed changes

Updates to the spin-weighted spherical harmonic transform utilities. These include putting in a public coefficients interface, creating an easy method of converting between libsharp coefficients and more traditional methods, and refactoring the transformation functions.



### Types of changes:

- [ ] New feature

### Component:

- [ ] Code

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Dependencies
 - [x] #1522

moxcodes committed:
Add swsh collocation point count functions

moxcodes committed:
Expand swsh transform utilities

- refactor common code to detail utilities
- move code to cpp where applicable

moxcodes committed:
Add SWSH Coefficients public interface

moxcodes commented:
Force-pushed a small change to get gcc-6 to build.

moxcodes commented:
This is no longer dependent, should be ready to review.
Cheers

hrueter commented on src/NumericalAlgorithms/Spectral/SwshTransformJob.hpp:
```diff
@@ -25,6 +26,12 @@
+namespace detail {
+// libsharp has an internal maximum number of transforms that is not in the
+// public interface, so we must hard-code its value here
+static const size_t MAX_LIBSHARP_TRANSFORMS = 100;
```
lower_snake_case

hrueter commented on src/NumericalAlgorithms/Spectral/SwshTransformJob.hpp:
```diff
@@ -208,6 +215,43 @@ class TransformJob {
+// point vector. This function will perform a conjugate on the collocations if
+// `positive_spin` is false, which will need to be reversed by the end of the
+// transform routine to restore the original values by the end of the transform
+// procedure.
```
We had trouble understanding this comment. Please refer to function arguments by name.

hrueter commented on src/NumericalAlgorithms/Spectral/SwshTransformJob.hpp:
```diff
@@ -227,66 +271,41 @@ void TransformJob<Spin, Representation, TagList>::execute_transform(
+    detail::assemble_libsharp_collocation_pointers(
+        make_not_null(&pre_transform_collocation_data),
+        make_not_null(&pre_transform_views),
+        make_not_null(&get(get<typename decltype(x)::type>(*input)).data()),
```
`using tag = tmpl::type_from<decltype(tag_v)>` and rename closure argument `auto x` to `auto tag_v`

hrueter commented on src/NumericalAlgorithms/Spectral/SwshTransformJob.hpp:
```diff
@@ -369,6 +373,85 @@ void TransformJob<Spin, Representation, TagList>::execute_inverse_transform(
 }
 
+// @{
+/*
```
`/*!`

hrueter commented on src/NumericalAlgorithms/Spectral/SwshTransformJob.hpp:
```diff
@@ -369,6 +373,85 @@ void TransformJob<Spin, Representation, TagList>::execute_inverse_transform(
+// @}
+
+// @{
+/*
```
`/*!`

hrueter commented on src/NumericalAlgorithms/Spectral/SwshTransformJob.hpp:
```diff
@@ -369,6 +373,85 @@ void TransformJob<Spin, Representation, TagList>::execute_inverse_transform(
+ * identical results but may help or hurt performance depending on the task.
+ * - `Spin`: The spin-weight of the quantity being transformed.
+ *
+ * The result is a set of libsharp-compatible coefficients.
```
The result is actually the collocation.

hrueter commented on src/NumericalAlgorithms/Spectral/SwshCoefficients.hpp:
```diff
@@ -345,13 +350,13 @@ template <int Spin>
 std::complex<double> libsharp_mode_to_goldberg_plus_m(
     const Coefficients::CoefficientsConstIterator& coefficient_iterator,
     const SpinWeighted<ComplexModalVector, Spin>& libsharp_modes,
-    const size_t angular_slice) noexcept;
```
Please move the Goldberg changes from the previous commit to this one.

hrueter commented on src/NumericalAlgorithms/Spectral/SwshCoefficients.cpp:
```diff
@@ -11,11 +11,12 @@
 
 namespace Spectral {
 namespace Swsh {
-namespace detail {
```
Why do you want to expose these libsharp specific implementation details? Ideally the public interface is independent of the backend library.

In particular we also think of functions with libsharp specific argument types and names, like the Goldberg functions you add in the fourth commit.

moxcodes replied:
It is necessary for the efficiency of calling code to be able to retrieve the format of data specific to the underlying library, for caching and dispatching to functions which use that data. Note also that this is not unlike the existing (ordinary) spherical harmonic code in `ApparentHorizons`, for which the returned coefficient data for public transforms is in a format specific to spherepack, and requires functionality similar to the iterator in this file to obtain mathematically familiar coefficients.

I don't disagree with the philosophy, but I'm afraid I don't see a way to entirely insulate calling code from the representation peculiarities without performing a lot of often unnecessary and costly conversions before delivering the coefficient data.

hrueter commented on src/NumericalAlgorithms/Spectral/SwshTransformJob.cpp:
```diff
@@ -154,17 +167,36 @@ SpinWeighted<ComplexModalVector, Spin> swsh_transform(
   return result_vector;
 }
 
+template <int Spin>
```
Instead of this overload use a default template parameter for the Representation.

moxcodes replied:
This one is troublesome for template parameter ordering. The point here is that the `Spin` parameter should always be inferred from the spin-weight of the argument, and that template defaults must follow non-defaulted template parameters. If the spin was first, however, it would try to match to the first specified template parameter in calling code. So, as far as I can tell, the template system does not permit from a single definition both the possibility of specifying only the representation and inferring the spin or having no specified parameters and defaulting the representation while inferring the spin.

Do you have a way around this template parameter ordering trouble?

hrueter replied:
I think you can actually have inferred template parameters coming after the defaulted template parameters.
By I think is, I wrote a test code doing this, which is compiling without complaints and is working. Furthermore I think there is no standard actually prohibiting this.

This is my test code:
```
#include <iostream>

template <int Spin>
class SpinWeighted {};

enum class ComplexRepresentation { Interleaved, RealsThenImags };

template <ComplexRepresentation Representation=ComplexRepresentation::Interleaved,
          int Spin>
void inverse_swsh_transform(
    SpinWeighted<Spin> input) noexcept{
    std::cout << "Spin: " << Spin << std::endl << "Representation: ";
    if(Representation==ComplexRepresentation::Interleaved)
        std::cout << "Interleaved";
    else
        std::cout << "RealsThenImags";
    std::cout << std::endl;
}
main() {
  SpinWeighted<1> in;
  inverse_swsh_transform<ComplexRepresentation::RealsThenImags>(in);
  inverse_swsh_transform(in);
}
```

moxcodes replied:
Aha, excellent. I must have misinterpreted a compiler error at some point when putting this together. Because the change effectively reverts one of the commits, I've squashed the change into that commit, and relabeled it with a `fixup` prefix to ultimately collapse into the other `SwshTransform` changes.

hrueter reviewed (changes requested):
group review by @hrueter (main) @nilsleiffischer 

Please fix the commit history so that the changes are in the commit they are intended for.

moxcodes committed:
Add Goldberg SWSH Coefficient routines

moxcodes commented:
fixups and responses posted for @hrueter.
Cheers!

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshCollocation.hpp:
```diff
@@ -23,6 +23,34 @@ number_of_swsh_collocation_points(const size_t l_max) noexcept {
 }
 
+/// \ingroup SwshGroup
+/// \brief Convenience function for determining the number of spin-weighted
```
shorten.

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshTransform.hpp:
```diff
@@ -208,13 +219,50 @@ class TransformJob {
+    gsl::not_null<ComplexDataVector*> vector, size_t l_max,
+    bool positive_spin) noexcept;
+
+// helper function for libsharp transforms which adds to a vector of pointers
```
`add`->`append`

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshTransform.hpp:
```diff
@@ -208,13 +219,50 @@ class TransformJob {
+// the set of pointers associated with angular views of the provided coefficient
+// vector. When working with the libsharp coefficient representation, note the
+// intricacies mentioned in the documentation for `TransformJob`.
+void assemble_libsharp_coefficient_pointers(
```
`assemble`->`append` or `append_to`

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshTransform.hpp:
```diff
@@ -208,13 +219,50 @@ class TransformJob {
+    gsl::not_null<ComplexModalVector*> vector, size_t l_max) noexcept;
+
+// helper function for the libsharp transforms which performs the actual
+// libsharp execution calls on a input and output set of pointers. This function
```
`a input` ->`an input`

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshTransform.hpp:
```diff
@@ -208,13 +219,50 @@ class TransformJob {
+    gsl::not_null<std::vector<std::complex<double>*>*> coefficient_data,
+    gsl::not_null<ComplexModalVector*> vector, size_t l_max) noexcept;
+
+// helper function for the libsharp transforms which performs the actual
```
shorten preamble

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshTransform.hpp:
```diff
@@ -227,66 +275,41 @@ void TransformJob<Spin, Representation, TagList>::execute_transform(
-  // pre_transform_collocation_data.
-  size_t num_transforms = 0;
+  // libsharp considers two arrays per transform when spin is not zero.
+  size_t num_transforms = (Spin == 0 ? 2 : 1) * number_of_radial_grid_points_ *
```
remove if unused, const if unmodified

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshTransform.cpp:
```diff
@@ -0,0 +1,335 @@
+       (num_transforms + MAX_LIBSHARP_TRANSFORMS - 1) / MAX_LIBSHARP_TRANSFORMS;
+       ++transform_block) {
+    if (transform_block < (num_transforms / MAX_LIBSHARP_TRANSFORMS)) {
+      // clang-tidy cppcoreguildelines-pro-bounds-pointer-arithmetic
```
`guidelines`

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshTransform.cpp:
```diff
@@ -0,0 +1,335 @@
+      sharp_execute(
+          jobtype, abs(spin),
+          coefficient_data->data() +  // NOLINT
+              (spin == 0 ? 1 : 2) * MAX_LIBSHARP_TRANSFORMS * transform_block,
```
explain unfortunate libsharp woes

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshCoefficients.hpp:
```diff
@@ -125,34 +129,141 @@ struct DestroySharpAlm {
+/*!
+ * \ingroup SwshGroup
+ * \brief A container for libsharp metadata for the spin-weighted spherical
+ * harmonics modal representation. This class also provides an iterator for
```
Second sentence out of brief

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshCoefficients.hpp:
```diff
@@ -125,34 +129,141 @@ struct DestroySharpAlm {
  public:
+  /// An iterator for easily traversing a libsharp-compatible spin-weighted
+  /// spherical harmonic modal representation.
+  /// The `operator*()` recovers the data associated with a pair of entries in
```
"...returns a `LibsharpCoefficientInfo`, which contains two offsets..."

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshCoefficients.hpp:
```diff
@@ -125,34 +129,141 @@ struct DestroySharpAlm {
-//   quantities, the m=0 modes always have vanishing imaginary component.
+}  // namespace detail
+
+// A container for reporting a single pair of modes in a libsharp-compatible
```
`container for reporting` -> `points to` (also make dox)

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshCoefficients.hpp:
```diff
@@ -125,34 +129,141 @@ struct DestroySharpAlm {
+    /// advance the iterator by one position (postfix)
+    // clang-tidy wants this to return a const iterator
+    CoefficientsConstIterator operator++(int) noexcept {  // NOLINT
+      if(l_ == l_max()) {
```
re-run clang-format

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshCoefficients.hpp:
```diff
@@ -125,34 +129,141 @@ struct DestroySharpAlm {
+    }
+    /// advance the iterator by one position (prefix)
+    CoefficientsConstIterator& operator++() noexcept {
+      if(l_ == l_max()) {
```
add debug `Assert` regarding appropriate `l` and `m` values

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshCoefficients.hpp:
```diff
@@ -125,34 +129,141 @@ struct DestroySharpAlm {
+        : m_{start_m}, l_{start_l}, coefficients_{coefficients} {}
+
+    /// Extracts the necessary information for accessing the current position in
+    /// a libsharp-compatible modal data structure
```
remove

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshCoefficients.hpp:
```diff
@@ -125,34 +129,141 @@ struct DestroySharpAlm {
+      return CoefficientsConstIterator(coefficients_, m_, l_);
+    };
+
+    // retreat the iterator by one position (prefix)
```
dox

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshCoefficients.hpp:
```diff
@@ -125,34 +129,141 @@ struct DestroySharpAlm {
+      } else {
+        ++l_;
+      }
+      return CoefficientsConstIterator(coefficients_, m_, l_);
```
this incorrectly acts as prefix (same with decrement)
simply discard postfix, there's not a good reason to use it.

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshCoefficients.hpp:
```diff
@@ -125,34 +129,141 @@ struct DestroySharpAlm {
+        --l_;
+      }
+      return CoefficientsConstIterator(coefficients_, m_, l_);
+    };
```
remove unnecessary semicolon.

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshCoefficients.hpp:
```diff
@@ -125,34 +129,141 @@ struct DestroySharpAlm {
+ * - due to the restriction of representing only the transform of real
+ *   quantities, the m=0 modes always have vanishing imaginary component.
+ */
 class Coefficients {
```
Rename this and the analogous `Collocation` to `CoefficientsMetadata` and `CollocationMetadata` for clarity.

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshCoefficients.hpp:
```diff
@@ -125,34 +129,141 @@ struct DestroySharpAlm {
+   private:
+    size_t m_;
+    size_t l_;
+    const Coefficients* const coefficients_;
```
just store `l_max` and make the equality more permissive.

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshCoefficients.hpp:
```diff
@@ -165,19 +276,48 @@ class Coefficients {
+  constexpr size_t l_max() const noexcept { return l_max_; }
+
+  constexpr size_t size() const noexcept {
+    return number_of_swsh_coefficients(l_max_);
```
2 * (plus comments and test)

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshCoefficients.hpp:
```diff
@@ -165,19 +276,48 @@ class Coefficients {
-// caching mechanism.
+/*!
+ * \ingroup SwshGroup
+ * \brief Generation function for obtaining a `Coefficients` object which is
```
changed `cached_coefficients` (also collocation similarly)

moxcodes commented on tests/Unit/NumericalAlgorithms/Spectral/Test_SwshCoefficients.cpp:
```diff
@@ -3,39 +3,49 @@
+  CHECK(precomputed_libsharp_lm.l_max() == l_max);
   CHECK(computed_coefficients.l_max() == l_max);
 
   sharp_alm_info* manual_sai;
```
`expected_sharp_alm_info`

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshCoefficients.hpp:
```diff
@@ -276,11 +281,9 @@ class Coefficients {
     return alm_info_.get();
   }
 
-  constexpr size_t l_max() const noexcept { return l_max_; }
```
put back :(

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshCoefficients.hpp:
```diff
@@ -276,11 +281,9 @@ class Coefficients {
+  size_t l_max() const noexcept { return l_max_; }
 
-  constexpr size_t size() const noexcept {
-    return number_of_swsh_coefficients(l_max_);
```
put back (and don't lose that factor of 2) and comments and test.

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshCoefficients.hpp:
```diff
@@ -319,5 +322,172 @@ class Coefficients {
+/*!
+ * \ingroup SwshGroup
+ * \brief Compute the mode coefficient for the convention of
+ * \cite Goldberg1966uu. See the documentation for `TransformJob` for complete
```
details to details

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshCoefficients.hpp:
```diff
@@ -319,5 +322,172 @@ class Coefficients {
+ * \param coefficient_iterator An iterator which stores an \f$l\f$ and \f$|m|\f$
+ * for the mode to be converted to the Goldberg form.
+ * \param libsharp_modes The libsharp-compatible modal storage to be accessed
+ * \param angular_slice The index of which angular slice is desired. Modes for
```
`radial_offset` (and below)

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshCoefficients.hpp:
```diff
@@ -319,5 +322,172 @@ class Coefficients {
+void libsharp_to_goldberg_modes(
+    const gsl::not_null<SpinWeighted<ComplexModalVector, Spin>*> goldberg_modes,
+    const SpinWeighted<ComplexModalVector, Spin>& libsharp_modes,
+    const size_t l_max) noexcept {
```
define in cpp instantiate

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshCoefficients.hpp:
```diff
@@ -319,5 +322,172 @@ class Coefficients {
+      libsharp_modes.data().size() /
+      (2 * Spectral::Swsh::number_of_swsh_coefficients(l_max));
+
+  if (UNLIKELY(goldberg_modes->data().size() !=
```
use `destructive_resize`

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshCoefficients.hpp:
```diff
@@ -319,5 +322,172 @@ class Coefficients {
+    for (auto coefficients_iterator = coefficients_metadata.begin();
+         coefficients_iterator != coefficients_metadata.end();
+         ++coefficients_iterator) {
+      goldberg_modes
```
comment or function

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshCoefficients.hpp:
```diff
@@ -319,5 +322,172 @@ class Coefficients {
+ */
+template <int Spin>
+std::complex<double> libsharp_mode_to_goldberg_plus_m(
+    const Coefficients::CoefficientsConstIterator& coefficient_iterator,
```
take info instead of iterator

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshCoefficients.hpp:
```diff
@@ -319,5 +322,172 @@ class Coefficients {
+    const size_t l_max) noexcept {
+  const size_t number_of_radial_grid_points =
+      libsharp_modes.data().size() /
+      (2 * Spectral::Swsh::number_of_swsh_coefficients(l_max));
```
change to `size_of_libsharp_coefficient_vector` (obviously changing the actual function).

moxcodes commented on tests/Unit/NumericalAlgorithms/Spectral/Test_SwshCoefficients.cpp:
```diff
@@ -93,6 +93,7 @@ void test_swsh_coefficients_class_interface() noexcept {
           manual_sai->mvstart[m_index]);
   }
 
+
```
remove

moxcodes commented on tests/Unit/NumericalAlgorithms/Spectral/Test_SwshCoefficients.cpp:
```diff
@@ -117,12 +118,148 @@ void test_swsh_coefficients_class_interface() noexcept {
+          make_not_null(&gen), make_not_null(&coefficient_distribution),
+          square(l_max + 1) * number_of_radial_points);
+
+  for (size_t i = 0; i < number_of_radial_points; ++i) {
```
comment

moxcodes commented on tests/Unit/NumericalAlgorithms/Spectral/Test_SwshCoefficients.cpp:
```diff
@@ -117,12 +118,148 @@ void test_swsh_coefficients_class_interface() noexcept {
+         ++coefficient_iter) {
+      auto goldberg_mode_plus_m =
+          libsharp_mode_to_goldberg_plus_m(coefficient_iter, test_modes, i);
+      auto alternative_mode_plus_m = libsharp_mode_to_goldberg(
```
comment

moxcodes commented on tests/Unit/NumericalAlgorithms/Spectral/Test_SwshCoefficients.cpp:
```diff
@@ -117,12 +118,148 @@ void test_swsh_coefficients_class_interface() noexcept {
+    }
+  }
+
+  auto inverse_transform_set_from_goldberg =
```
maybe const

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshCoefficients.cpp:
```diff
@@ -54,5 +58,163 @@ const Coefficients& precomputed_coefficients(const size_t l_max) noexcept {
+    const size_t l, const int m, const size_t l_max,
+    const SpinWeighted<ComplexModalVector, Spin>& libsharp_modes,
+    const size_t angular_slice) noexcept {
+  auto& libsharp_harmonic_lm = precomputed_coefficients(l_max);
```
const and rename

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshCoefficients.cpp:
```diff
@@ -54,5 +58,163 @@ const Coefficients& precomputed_coefficients(const size_t l_max) noexcept {
+      sharp_swsh_sign(Spin, (*coefficient_iterator).m, false) *
+          sharp_swsh_sign(Spin, -static_cast<int>((*coefficient_iterator).m),
+                          true));
+  auto i = std::complex<double>(0.0, 1.0);
```
const

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshCoefficients.cpp:
```diff
@@ -54,5 +58,163 @@ const Coefficients& precomputed_coefficients(const size_t l_max) noexcept {
+    const size_t angular_slice,
+    const std::complex<double> goldberg_plus_m_mode_value,
+    const std::complex<double> goldberg_minus_m_mode_value) noexcept {
+  auto& libsharp_harmonic_lm = precomputed_coefficients(l_max);
```
const rename

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshTransform.hpp:
```diff
@@ -221,11 +221,12 @@ class TransformJob {
 
 namespace detail {
 // helper function for libsharp transforms which adds to a vector of pointers
```
`helper function for libsharp transforms which adds` -> `appends`

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshTransform.hpp:
```diff
@@ -221,11 +221,12 @@ class TransformJob {
-// transform routine to restore the original values by the end of the transform
-// procedure.
+// `collocation_data` the set of pointers associated with angular views of the
+// provided collocation data `vector`. The views are stored in
```
`The associated views are appended into`

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshTransform.hpp:
```diff
@@ -221,11 +221,12 @@ class TransformJob {
-// procedure.
+// `collocation_data` the set of pointers associated with angular views of the
+// provided collocation data `vector`. The views are stored in
+// `collocation_views`. This function will perform a conjugate on the
```
"if `positive_spin` is false, this function will conjugate the views, and therefore potentially conjugate (depending on `Representation`) the input data `vectors`. This behavior is chosen to avoid frequent copies of the input data `vector` for optimization. The conjugation must be undone after the call to libsharp transforms using `FUNCTION NAME` in order to return the input data to its original state."
and add `FUNCTION NAME`

moxcodes reviewed (commented):
[request changes]
group code review with @fmahebert @markscheel .

moxcodes committed:
rename swsh collocation metadata utilities

moxcodes committed:
Add SpinWeighted destructive_resize

moxcodes commented:
@fmahebert @markscheel fixups posted
Cheers!

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshCollocation.hpp:
```diff
@@ -24,9 +24,8 @@ number_of_swsh_collocation_points(const size_t l_max) noexcept {
-/// spherical harmonic collocation values in the \f$\theta\f$ direction that are
-/// stored for a given `l_max` for a libsharp-compatible set of collocation
+/// \brief Returns the number of spin-weighted spherical harmonic collocation
+/// values in the \f$\theta\f$  for a libsharp-compatible set of collocation
```
remove `the` and extra whitespace

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshCollocation.hpp:
```diff
@@ -38,9 +37,8 @@ number_of_swsh_theta_collocation_points(const size_t l_max) noexcept {
-/// spherical harmonic collocation values in the \f$\phi\f$ direction that are
-/// stored for a given `l_max` for a libsharp-compatible set of collocation
+/// \brief Returns the number of spin-weighted spherical harmonic collocation
+/// values in the \f$\phi\f$ for a libsharp-compatible set of collocation
```
remove the

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshTransform.hpp:
```diff
@@ -221,13 +221,15 @@ class TransformJob {
+// conjugate (depending on `Representation`) the input data `vectors`. This
+// behavior is chosen to avoid frequent copies of the input data `vector` for
+// optimization. The conjugation must be undone after the call to libsharp
+// transforms using `conjugate_swsh_transform_data_views` to restore the input
```
`conjugate_views`

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshTransform.hpp:
```diff
@@ -236,10 +238,25 @@ void append_libsharp_collocation_pointers(
+    gsl::not_null<std::vector<ComplexDataView<Representation>>*>
+        collocation_views) noexcept {
+  if (Spin < 0) {
+    // fix the conjugation performed prior to transformation.
```
unnecessary

moxcodes commented on src/NumericalAlgorithms/Spectral/SwshTransform.hpp:
```diff
@@ -236,10 +238,25 @@ void append_libsharp_collocation_pointers(
-// vector. When working with the libsharp coefficient representation, note the
-// intricacies mentioned in the documentation for `TransformJob`.
+// Perform a conjugation on a vector of `ComplexDataView`s if `Spin` < 0. This
+// is used for fixing the operation described in the code comment for
```
`for undoing the conjugation`

moxcodes reviewed (commented):
A few more comment changes group code review with @markscheel @fmahebert (and @fmahebert says squash)

moxcodes commented:
squashed (along with a trivial clang-tidy fix).
Thanks!

moxcodes commented:
@sxs-collaboration/spectre-core-devs This should be ready for a last look.
Thanks!

nilsdeppe commented on src/NumericalAlgorithms/Spectral/SwshTransform.cpp:
```diff
@@ -0,0 +1,295 @@
+       (num_transforms + max_libsharp_transforms - 1) / max_libsharp_transforms;
+       ++transform_block) {
+    if (transform_block < (num_transforms / max_libsharp_transforms)) {
+      // spin ternaries below are used because libsharp considers two arrays per
```
It seems to me that this would be easier to read if you factored out the ternaries into a `spin_factor` (or some other name). What do you think?

nilsdeppe commented on src/NumericalAlgorithms/Spectral/SwshTransform.cpp:
```diff
@@ -0,0 +1,295 @@
+  std::vector<double*> pre_transform_collocation_data;
+  pre_transform_collocation_data.reserve(2 * number_of_radial_points);
+
+  const auto collocation_metadata =
```
It looks like this will be a pointer, in which case `const auto*` would make the code a lot easier to read :)

nilsdeppe commented on src/NumericalAlgorithms/Spectral/SwshCoefficients.hpp:
```diff
@@ -125,59 +133,183 @@ struct DestroySharpAlm {
+          l_max_, l_, m_};
+    }
+    /// advance the iterator by one position (prefix)
+    CoefficientsConstIterator& operator++() noexcept {
```
Can you add the `operator++(int)` and `operator--(int)` variants as well for completeness? I can see someone (like myself) expecting to be able to do both `++it` and `it++`

moxcodes replied:
@markscheel and @fmahebert suggested removing the postfix operators I originally had in this PR under the philosophy that because they cause copies, it's better to present only the interface that one would suggest for current use-cases. If you feel strongly about following the standard, though, I can put them back.

nilsdeppe replied:
If copying 3 ints is what kills us I'll celebrate :D I'd rather have a conforming interface. There are cases where the postfix is actually what is preferred, i.e. one wants the copy. I know @fmahebert doesn't like people doing that, but I think it makes more sense to have a predictable interface.

fmahebert replied:
Fine 😛

nilsdeppe commented on src/NumericalAlgorithms/Spectral/SwshCoefficients.hpp:
```diff
@@ -125,59 +133,183 @@ struct DestroySharpAlm {
+
+  // @{
+  /// \brief Get a bidirectional iterator to the start of the series of modes.
+  CoefficientsMetadata::CoefficientsConstIterator begin() const noexcept {
```
To make sure I understand: these iterators leave the underlying data immutable, right? So I could not change the value of a particular coefficient using these iterators.

moxcodes replied:
Note that these iterators dereference to objects which just contain l, m, and offests into a vector. So, the mutability of the underlying data is determined by the vector you use the offset to index.

nilsdeppe replied:
Hmm, okay so this is more like a `CoefficientsIndexIterator` then. I'd prefer a name change that makes it clear you are not going to get a modal coefficient when dereferencing the iterator. What do you think?

moxcodes replied:
Yeah, I'm happy making it `CoefficientsIndexIterator`

nilsdeppe commented on src/NumericalAlgorithms/Spectral/SwshCoefficients.hpp:
```diff
@@ -311,5 +316,158 @@ class CoefficientsMetadata {
+ * determine the representation details.
+ */
+template <int Spin>
+std::complex<double> libsharp_mode_to_goldberg(
```
to be consistent with the rest of the "spectral" code, probably `libsharp_mode_to_goldberg` and `goldberg_to_libsharp_mode` would be the preferred names (as opposed to a `from`). For example, the orthogonal polynomials use `modal_to_nodal` and `nodal_to_modal`

nilsdeppe reviewed (commented):
Just a few suggestions and ideas I'd like to discuss :)

moxcodes commented:
@nilsdeppe : we (@fmahebert , @markscheel  and I) have had a discussion regarding the transform utilities here. There is the trouble that the 'input' to the transform utility has the possibility of getting conjugated in place during the transform, but the conjugation is reversed by the end. This prevents the input from being const, which causes a proliferation of needing to pass a great deal by pointer and flagging things as mutated when they won't be changed by the end of the transform. This ends up happening a lot in CCE code.

The (slightly scary) suggestion is to actually take the argument by const and const-cast it away to do the conjugation if necessary. Because none of the public interfaces fail to undo the conjugation by the end of the functions, the inputs are 'morally const', and in the charm framework, if we understand correctly, no two actions will ever execute simultaneously on the same component. So, it should be 'safe' provided all of the data passed to a transform is in a `DataBox`.

So, what do you think: should we keep it as a pass-by-pointer because of the intermediate modification and tolerate 'extra' pass-by-pointer etc. on later code, or should we take the input by const, const-cast and carefully document the behavior to keep the interface in a 'tidy' form?

moxcodes commented:
fixups posted for the last couple of review requests.
Cheers

nilsdeppe commented:
@moxcodes so I've done this const-cast quite a bit in the subcell code (and I think also in things that are already in develop) when I need to use `DataVector`'s referencing functionality to get at the expression templates. In these cases as well the input is taken be `const` since from the interface's perspective the data remains unchanged. I'm in favor of this approach because I think it's conceptually clearer for the user of the interface that afterward the data is the same, even though temporarily in the middle it might be changed. The only way two things can be run on a parallel component at once is if the component is a node group, you're using a threaded action, and not using a lock in the place where it is needed. This seems like an unlikely scenario for the CCE code (and anything else that uses these tools). I'm in favor of the `const_cast` and cleaner interface :)

nilsdeppe commented:
fixups look good to me. I'm happy for you to squash! I think the only thing I didn't see (and maybe I missed it) is `operator++(int)`. Feel free to squash that immediately as well!

moxcodes committed:
Adjust transforms to const inputs

- const_casts used internally, but input is guaranteed to return to original
  state by the end of the transforms, so is 'morally const'

moxcodes commented:
@nilsdeppe @fmahebert I've squashed in the fixups (and put the postfix operators in that I missed bringing back in). I've added a commit to the end of the series that makes the switch to const inputs.
Thanks

nilsdeppe reviewed (dismissed):
LGTM!  :+1:

moxcodes commented:
sorry, one more squash 'cause clang-tidy didn't like the const casts.
